src/unity/python/turicreate/toolkits/drawing_recognition/_sframe_loader.py

```diff
@@ -0,0 +1,282 @@
+# -*- coding: utf-8 -*-
+# Copyright © 2019 Apple Inc. All rights reserved.
+#
+# Use of this source code is governed by a BSD-3-clause license that can
+# be found in the LICENSE.txt file or at https://opensource.org/licenses/BSD-3-Clause
+from __future__ import print_function as _
+from __future__ import division as _
+from __future__ import absolute_import as _
+import numpy as _np
+import turicreate as _tc
+import mxnet as _mx
+from six.moves.queue import Queue as _Queue
+from threading import Thread as _Thread
+from turicreate.toolkits._main import ToolkitError as _ToolkitError
+
+_TMP_COL_RANDOM_ORDER = '_random_order'
+
+# Encapsulates an SFrame, iterating over each row and returning an
+# (image, label, index) tuple.
+class _SFrameDataSource:
+    def __init__(self, sframe, feature_column, target_column,
+                 load_labels=True, shuffle=True, samples=None):
+        self.target_column = target_column
+        self.feature_column = feature_column
+        self.load_labels = load_labels
+        self.shuffle = shuffle
+        self.num_samples = samples
+        self.cur_sample = 0
+
+        # Make shallow copy, so that temporary columns do not change input
+        self.sframe = sframe.copy()
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        return self._next()
+
+    def next(self):
+        return self._next()
+
+    def _next(self):
+        if self.cur_sample == self.num_samples:
+            raise StopIteration
+
+        # If we're about to begin a new epoch, shuffle the SFrame if requested.
+        row_index = self.cur_sample % len(self.sframe)
+        if row_index == 0 and self.cur_sample > 0 and self.shuffle:
+            self.sframe[_TMP_COL_RANDOM_ORDER] = _np.random.uniform(size=len(self.sframe))
+            self.sframe = self.sframe.sort(_TMP_COL_RANDOM_ORDER)
+        self.cur_sample += 1
+
+        # Copy the image data for this row into a NumPy array.
+        row = self.sframe[row_index]
+
+        drawing_feature = row[self.feature_column]
+        is_stroke_input = (type(drawing_feature) != _tc.Image)
+        if is_stroke_input:
+            pixels_01 = drawing_feature
+        else:    
+            image = _tc.image_analysis.resize(drawing_feature, 28, 28, 1)
+            pixels_01 = image.pixel_data.reshape(1, 28, 28) / 255.
+        
+        if self.load_labels:
+            label = row[self.target_column]
+            if label is None:
+                label = []
+            elif type(label) is dict:
+                label = [label]
+        else:
+            label = None
+
+        return pixels_01, label, row_index
+
+    def reset(self):
+        self.cur_sample = 0
+
+
+# A wrapper around _SFrameDataSource that uses a dedicated worker thread for
+# performing SFrame operations.
+class _SFrameAsyncDataSource:
```

Comment:
Does the async data prep actually help? Might be worth seeing if the performance is fine without it. This adds significant complexity that seemed needed for object detection, but I'm not convinced we need it here, especially if we can't factor it out and share it among toolkit data iterators

Reply:
Removed Async Data Prep.